core/src/main/java/org/bitcoinj/base/BitcoinNetwork.java

```diff
@@ -182,6 +182,51 @@ public boolean exceedsMaxMoney(Monetary amount) {
         }
     }
 
+    /**
+     * Check if an address is valid on this network.
+     * This is meant to be used as a precondition for a method or function that expects a valid address. If
+     * you are validating addresses provided externally, you probably want to use
+     * {@link #isValidAddress(Address)} to handle errors more gracefully.
+     * @param address Address to validate
+     * @return The unmodified address if valid on this network
+     * @throws IllegalArgumentException if address not valid on this network
+     */
+    public Address checkAddress(Address address) throws IllegalArgumentException {
+        if (!isValidAddress(address)) {
+            throw new IllegalArgumentException(String.format("Address %s not valid on network %s", address, this));
+        }
+        return address;
+    }
+
+    /**
+     * Is address valid for this network.
+     * @param address Address to validate
+     * @return {@code true} if valid on this network, {@code false} otherwise
+     */
+    public boolean isValidAddress(Address address) {
+        boolean valid;
+        switch (this) {
+            case MAINNET:
+                valid = address.network() == MAINNET;
+                break;
+            case TESTNET:
+            case SIGNET:
+                valid = address.network() == TESTNET;
+                break;
+            case REGTEST:
+                if (address instanceof LegacyAddress) {
+                    valid = ((LegacyAddress) address).network == TESTNET;
+                } else {
+                    valid = address.network() == REGTEST;
+                }
```

Comment:
What's the backstory behind this logic?

Reply:
I'm assuming you are asking me to add the "backstory" to the comments? I think that is a good idea. I'll add it.

Reply:
Fixed.